### PR TITLE
Rename prepper to processor

### LINKED ISSUE #311: Rename Prepper to Processor
**Is your feature request related to a problem? Please describe.**

Data Prepper currently supports `prepper` plugins. This term is not entirely clear. Additionally, it is somewhat ambiguous with Data Prepper.

**Describe the solution you'd like**

Rename `prepper` to `processor`.

This can be done in a phased approach. First, we will add `processor` as a supported name within the pipeline YAML configuration. We will continue to support either `processor` or `prepper` in the YAML. However, only one may be present per pipeline.

In a major release version of Data Prepper we will remove `prepper` so that only `processor` is allowed.

Additionally, we will add a new interface `Processor` with the same signature as `Prepper`. The `Prepper` interface will be made to inherit from `Processor` and will be deprecated. In a major release of Data Prepper, we will remove the `Prepper` interface.

### PR DESCRIPTION
**Pull request by sbayer55**
### Description
Removes references to Preppers in data-prepper-core and docs except where required for plugin compatibility.
 
### Issues Resolved
closes #311 
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by cmanning09 — COMMENTED** [review]
PR details links to a previously abandoned PR. Please update the correct issue in the PR details.

File: data-prepper-core/src/main/java/com/amazon/dataprepper/parser/model/PipelineConfiguration.java [hunk 1]
```diff
@@ -34,7 +34,7 @@
             final List<Map.Entry<String, Map<String, Object>>> preppers,
             final List<Map.Entry<String, Map<String, Object>>> processors) {
         if (preppers != null) {
-            LOG.warn("prepper configurations are deprecated, processor configurations will be required in 1.2");
+            LOG.warn("Prepper configurations are deprecated, processor configurations will be required in a future version");
```
[the remark this answers is not in the document]
Comment (sbayer55) [reply data-prepper-core/src/main/java/com/amazon/dataprepper/parser/model/PipelineConfiguration.java]: Updated warning to refer to version 2.0

**Review by dapowers87 — APPROVED** [review]

File: data-prepper-core/src/main/java/com/amazon/dataprepper/parser/PipelineParser.java [hunk 2]
```diff
@@ -8,8 +8,10 @@
 import com.amazon.dataprepper.model.annotations.SingleThread;
 import com.amazon.dataprepper.model.buffer.Buffer;
 import com.amazon.dataprepper.model.configuration.PluginSetting;
+import com.amazon.dataprepper.model.plugin.NoPluginFoundException;
 import com.amazon.dataprepper.model.plugin.PluginFactory;
 import com.amazon.dataprepper.model.prepper.Prepper;
```
Comment (dinujoh) [inline data-prepper-core/src/main/java/com/amazon/dataprepper/parser/PipelineParser.java:13]: Do we need Prepper import ?

Comment (dinujoh) [reply data-prepper-core/src/main/java/com/amazon/dataprepper/parser/PipelineParser.java:13]: Looks like we need this as fallback when processor type plugin is not found.

**Review by cmanning09 — DISMISSED** [review]
> PR details links to a previously abandoned PR. Please update the correct issue in the PR details.

#652  is still referenced in the details. Please update to a valid issue.

File: data-prepper-core/src/test/java/com/amazon/dataprepper/pipeline/PipelineTests.java [hunk 3]
```diff
@@ -58,23 +58,23 @@ public void testPipelineState() {
     }
 
     @Test
-    public void testPipelineStateWithPrepper() {
+    public void testPipelineStateWithProcessor() {
```
Comment (dlvenable) [inline data-prepper-core/src/test/java/com/amazon/dataprepper/pipeline/PipelineTests.java:61]: I think you should keep the existing test (`testPipelineStateWithPrepper`) and add the new one. This way the tests are testing both a `Prepper` and a `Processor`.

File: data-prepper-core/src/main/java/com/amazon/dataprepper/parser/PipelineParser.java [hunk 4]
```diff
@@ -115,11 +117,26 @@ private void buildPipelineFromConfiguration(
 
     }
 
-    private List<Prepper> newPreppers(final PluginSetting pluginSetting) {
-        return pluginFactory.loadPlugins(Prepper.class, pluginSetting,
+    private List<Processor> newProcessor(final PluginSetting pluginSetting) {
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/com/amazon/dataprepper/parser/PipelineParser.java:120]: This looks good. Thanks for adding it.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-core/src/main/java/com/amazon/dataprepper/parser/PipelineParser.java (hunk 2 above)
Comment (sbayer55) [reply data-prepper-core/src/main/java/com/amazon/dataprepper/parser/PipelineParser.java:13]: Yes. The current implementation of the @DataPrepperPlugin annotation and how plugin instances are created requires an exact class match on the pluginType property. Until all Prepper plugins have migrated to the Processor class plugin type the newProcessor will need to be able to fallback to casting a new Prepper instance to a Processor.

Alternative options would be:
1. Update all plugin annotations within this PR
2. Rework how plugin instances are loaded
3. Hard code a conversion of prepper to processor within the DefaultPluginFactory

Fallback process: https://github.com/sbayer55/data-prepper/blob/rename-prepper-to-processor/data-prepper-core/src/main/java/com/amazon/dataprepper/parser/PipelineParser.java#L127

Plugin factory class matching: https://github.com/sbayer55/data-prepper/blob/rename-prepper-to-processor/data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/DefaultPluginFactory.java#L104

File: data-prepper-core/src/test/java/com/amazon/dataprepper/pipeline/PipelineTests.java (hunk 3 above)
Comment (sbayer55) [reply data-prepper-core/src/test/java/com/amazon/dataprepper/pipeline/PipelineTests.java:82]: I added a testPipelineStateWithPrepper, please confirm my implementation was what you intended. https://github.com/sbayer55/data-prepper/blob/rename-prepper-to-processor/data-prepper-core/src/test/java/com/amazon/dataprepper/pipeline/PipelineTests.java#L62

Comment (dlvenable) [reply data-prepper-core/src/test/java/com/amazon/dataprepper/pipeline/PipelineTests.java:82]: Yes, it is. Thank you.

**Review by dlvenable — APPROVED** [review]

**Review by cmanning09 — APPROVED** [review]

**Pull request merged**